### PR TITLE
feat(kda): add KDA attention components

### PR DESCRIPTION
**Pull request by MokusMokun**
## Summary

Adds the KDA (Kimi Delta Attention) inference components for sglang-jax, enabling hybrid linear-attention model serving (e.g. Kimi-Linear, #937). Includes DP support via `shard_map`.

## Changes

**Kernels** (`srt/kernels/kda/`)
- `kda.py` — chunked Pallas forward (gate cumsum → intra-chunk delta-rule solve → inter-chunk state propagation → output computation), varlen alignment helpers
- `naive.py` — step-by-step recurrent reference, used as decode path and test baseline

**Layers** (`srt/layers/attention/linear/`)
- `short_convolution.py` — stateless depthwise causal conv1d with EXTEND (cu_seqlens-aware) and DECODE (rolling cache) paths
- `kda_backend.py` — `KDAAttnBackend` extending `LinearRecurrentAttnBackend`: conv → L2 norm → recurrent kernel dispatch, conv state pack/unpack, pool gather/scatter via `shard_map` for DP

**Tests** (`test/`)
- `test_short_conv.py` — 24 tests, conv correctness vs `nnx.Conv` baseline (fp32 + bf16)
- `test_kda_attention.py` — 9 backend tests at TP=4 (rtol/atol = 2e-2 / 1e-2)
- `test_kda_attention_dp.py` — 9 DP tests covering DP=4/TP=1 and DP=2/TP=2 (sparse ranks, unbalanced batches, mixed `has_initial_state`, multi-round prefill→decode)

## Test plan

- [x] `pytest python/sgl_jax/test/test_short_conv.py` → 24/24 (TPU v6e-4)
- [x] `pytest python/sgl_jax/test/test_kda_attention.py` → 9/9 (TPU v6e-4)
- [x] `pytest python/sgl_jax/test/test_kda_attention_dp.py` → 9/9 (TPU v6e-4)

## References

- RFC: #1038
- Issue: #937
- Hybrid memory pool: #1034
- Linear attention dispatch: #1041

### CODE REVIEW

File: python/sgl_jax/test/test_kda_attention.py [hunk 1]
```diff
@@ -0,0 +1,584 @@
...
+from sgl_jax.srt.managers.schedule_batch import ModelWorkerBatch
+from sgl_jax.srt.mem_cache.recurrent_state_pool import RecurrentStatePool
+from sgl_jax.srt.model_executor.forward_batch_info import ForwardBatch, ForwardMode
+from sgl_jax.srt.utils.mesh_utils import create_device_mesh
+
+mesh = create_device_mesh(ici_parallelism=[1, -1], dcn_parallelism=[1, 1])
+jax.sharding.set_mesh(mesh)
+
+
+def _scaled_randn(rng: np.random.Generator, shape, scale: float = 0.1) -> np.ndarray:
```
Comment (MokusMokun) [inline python/sgl_jax/test/test_kda_attention.py:23]: reasoning behind scale = 0.1: https://github.com/sgl-project/sglang-jax/issues/1038#issuecomment-4415313208

File: python/sgl_jax/test/test_kda_attention_dp.py [hunk 2]
```diff
@@ -0,0 +1,853 @@
...
+
+
+def _scaled_randn(rng: np.random.Generator, shape, scale: float = 0.1) -> np.ndarray:
+    # @Scale: 0.1 for kernel inputs/initial states to keep bf16 mantissa precision
+    #   in the delta-rule recurrence (S = S*exp(g) + beta*k*(v - k^T S))
+    #   without it, round-off error will accumulate across K-dim reductions and T recurrent steps.
+    return rng.standard_normal(shape).astype(np.float32) * scale
+
+
+# --- Reference baselines (verbatim from test_kda_attention.py).
```
Comment (MokusMokun) [inline python/sgl_jax/test/test_kda_attention_dp.py]: any better apporach to implement the reference baseline aside from verbatim ?
(as the `make_nnx_depthwise_conv`, `ref_short_convolution` , `ref_kda_attention`)

File: python/sgl_jax/test/test_kda_attention.py [hunk 3]
```diff
@@ -0,0 +1,584 @@
...
+        ssm, conv = self.random_states(batch_size=1)
+        self.run_test(
+            "prefill",
+            [64],
+            all_have_initial_state=True,
+            initial_ssm_state=ssm,
+            initial_conv_state=conv,
+        )
+
+    def test_extend_state_output(self):
```
Comment (MokusMokun) [inline python/sgl_jax/test/test_kda_attention.py]: deleted

File: python/sgl_jax/test/test_kda_attention.py (hunk 1 above)
Comment (aolemila) [reply python/sgl_jax/test/test_kda_attention.py:23]: For https://github.com/sgl-project/sglang-jax/issues/1038#issuecomment-4415313208, I still do not get your point. Let's do a discussion. Can I regard scale=0.1 is used to shorten the range of values and let them pass the specified tolerance?

File: python/sgl_jax/test/test_kda_attention.py [hunk 4]
```diff
@@ -0,0 +1,584 @@
...
+):
+    assert mode in ("prefill", "decode")
+    forward_mode = ForwardMode.EXTEND if mode == "prefill" else ForwardMode.DECODE
+    batch_size = len(seq_lens)
+    total_tokens = sum(seq_lens)
+    hidden = num_heads * head_dim
+    conv_sharding = NamedSharding(test_mesh, P("tensor", None))
+    hidden_sharding = NamedSharding(test_mesh, P("data", "tensor"))
+    head_sharding = NamedSharding(test_mesh, P("data", "tensor"))
+    param_head_sharding = NamedSharding(test_mesh, P("tensor", None))
```
Comment (aolemila) [inline python/sgl_jax/test/test_kda_attention.py:244]: Why do you add data in test_kda_attention.py? In my mind, this test does not involve data sharding. Maybe I misunderstand.

File: python/sgl_jax/test/test_kda_attention.py [hunk 5]
```diff
@@ -0,0 +1,584 @@
...
+    write_initial_state(pool, layer_id, recurrent_indices, initial_ssm_state, initial_conv_state)
+
+    seq_lens_np = np.asarray(seq_lens, dtype=np.int32)
+    input_ids = np.arange(total_tokens, dtype=np.int32)
+    positions = np.arange(total_tokens, dtype=np.int32)
+    out_cache_loc = np.arange(total_tokens, dtype=np.int32)
+    req_pool_indices = np.arange(batch_size, dtype=np.int32)
+    extend_seq_lens = seq_lens_np if mode == "prefill" else None
+    extend_prefix_lens = np.zeros(batch_size, dtype=np.int32) if mode == "prefill" else None
+    has_initial_state_np = np.full(batch_size, all_have_initial_state, dtype=np.bool_)
```
Comment (aolemila) [inline python/sgl_jax/test/test_kda_attention.py]: Could you add the case of mixing the value of have_initial_state to simulate the mixing scheduling for chunk requests and decode requests?

File: python/sgl_jax/test/test_kda_attention.py [hunk 6]
```diff
@@ -0,0 +1,584 @@
...
+        real_bs=batch_size,
+        real_bs_per_dp=[batch_size],
+        dp_size=1,
+        per_dp_bs_size=batch_size,
+        spec_info=None,
+        recurrent_indices=recurrent_indices,
+        has_initial_state=has_initial_state_np,
+    )
+
+    fb = ForwardBatch(
```
Comment (aolemila) [inline python/sgl_jax/test/test_kda_attention.py]: I recommend you using the `ForwardBatch.init_new()` generates the fb. `model_runner` is required but only mesh is used in this function.

File: python/sgl_jax/test/test_kda_attention.py [hunk 7]
```diff
@@ -0,0 +1,584 @@
...
+
+    # Replicate sharded tensors to host so per-seq slicing works under
+    # JAX's explicit-sharding mode (slicing a `data`-sharded axis would
+    # otherwise raise ShardingTypeError on the gather).
+    def unshard_for_ref(x):
+        return jnp.asarray(np.asarray(x), dtype=x.dtype)
+
+    g = unshard_for_ref(g)
+    b = unshard_for_ref(b)
+    initial_ssm_state = unshard_for_ref(initial_ssm_state)
```
Comment (aolemila) [inline python/sgl_jax/test/test_kda_attention.py]: this looks like that you remove the tensor sharding

File: python/sgl_jax/test/test_kda_attention.py [hunk 8]
```diff
@@ -0,0 +1,584 @@
...
+        recurrent_indices=jnp.asarray(recurrent_indices),
+    )
+    fb.attn_backend.forward_metadata = backend.get_forward_metadata(mwb)
+
+    baseline_ssm = (
+        initial_ssm_state if all_have_initial_state else jnp.zeros_like(initial_ssm_state)
+    )
+    baseline_conv = (
+        initial_conv_state if all_have_initial_state else jnp.zeros_like(initial_conv_state)
+    )
```
Comment (aolemila) [inline python/sgl_jax/test/test_kda_attention.py:376]: You have set the baseline_xxx with zeros_like(like clear_slots) before calling attention backend. So it means you can not test the function of clear_slots in backend. If I misunderstood, please let me know.

File: python/sgl_jax/test/test_kda_attention.py (hunk 4 above)
Comment (MokusMokun) [reply python/sgl_jax/test/test_kda_attention.py:244]: it a mandatory specification, limited by the shard_map design of `kda_backend.py`

File: python/sgl_jax/test/test_kda_attention.py (hunk 5 above)
Comment (MokusMokun) [reply python/sgl_jax/test/test_kda_attention.py]: added
```
def test_decode_mixed_new_continuing(self):
        ssm, conv = self.random_states(batch_size=4)
        self.run_test(
            "decode",
            [1, 1, 1, 1],
            all_have_initial_state=[True, False, True, False],
            initial_ssm_state=ssm,
            initial_conv_state=conv,
        )
```

File: python/sgl_jax/test/test_kda_attention.py (hunk 6 above)
Comment (MokusMokun) [reply python/sgl_jax/test/test_kda_attention.py]: refactored

File: python/sgl_jax/test/test_kda_attention.py [hunk 9]
```diff
@@ -0,0 +1,584 @@
...
+                self.atol,
+            )
+            np.testing.assert_allclose(
+                np.asarray(gather_conv(pool, conv_buffer_list[0], recurrent_indices)),
+                np.asarray(conv_ref),
+                self.rtol,
+                self.atol,
+            )
+            pool.replace_buffer(([recurrent_buffer], [conv_buffer_list]))
+
```
Comment (aolemila) [inline python/sgl_jax/test/test_kda_attention.py:616]: For this test, it lacks of the comparison of output and states between tp=1 and tp=4.

File: python/sgl_jax/test/test_kda_attention_dp.py [hunk 10]
```diff
@@ -0,0 +1,853 @@
...
+        extend_prefix_lens=(
+            jnp.asarray(extend_prefix_lens_cpu) if extend_prefix_lens_cpu is not None else None
+        ),
+        extend_seq_lens=(
+            jnp.asarray(extend_seq_lens_cpu) if extend_seq_lens_cpu is not None else None
+        ),
+        spec_info=None,
+        recurrent_indices=jnp.asarray(recurrent_indices_cpu),
+    )
+    fb.attn_backend.forward_metadata = backend.get_forward_metadata(mwb)
```
Comment (aolemila) [inline python/sgl_jax/test/test_kda_attention_dp.py:441]: the same to the above.

File: python/sgl_jax/test/test_kda_attention_dp.py [hunk 11]
```diff
@@ -0,0 +1,853 @@
...
+                continue
+            valid = sum(info["seq_lens"]) if is_prefill else len(info["seq_lens"])
+            offset = dp_rank * per_dp_token_padding if is_prefill else dp_rank * per_dp_bs_padding
+            np.testing.assert_allclose(
+                actual_np[offset : offset + valid],
+                expected[offset : offset + valid],
+                rtol=self.rtol,
+                atol=self.atol,
+                err_msg=f"DP rank {dp_rank} output mismatch",
+            )
```
Comment (aolemila) [inline python/sgl_jax/test/test_kda_attention_dp.py:645]: Lack of states comparison

File: python/sgl_jax/test/test_kda_attention_dp.py [hunk 12]
```diff
@@ -0,0 +1,853 @@
...
+                    continue
+                num_seqs = len(info_d["seq_lens"])
+                offset = dp_rank * per_dp_bs_padding_d
+                np.testing.assert_allclose(
+                    actual_d_np[offset : offset + num_seqs],
+                    expected_d[offset : offset + num_seqs],
+                    rtol=self.rtol,
+                    atol=self.atol,
+                    err_msg=f"decode round {step}: rank {dp_rank}",
+                )
```
Comment (aolemila) [inline python/sgl_jax/test/test_kda_attention_dp.py:838]: lack of states comparison

File: python/sgl_jax/test/test_kda_attention_dp.py [hunk 13]
```diff
@@ -0,0 +1,853 @@
...
+
```
Comment (aolemila) [inline python/sgl_jax/test/test_kda_attention_dp.py:839]: In this test file, you have the same problem in the above test file. You missing to test the clear_slots function in attention_backend.

File: python/sgl_jax/test/test_kda_attention_dp.py [hunk 14]
```diff
@@ -0,0 +1,853 @@
...
+
+if __name__ == "__main__":
+    unittest.main()
```
Comment (aolemila) [inline python/sgl_jax/test/test_kda_attention_dp.py:854]: In this test, only one tp>1 case exists. Could you add more cases which only change of tp and dp combination. I suggest creating a new file like test_kda_attention_dp_2_tp4.py to change the tp and dp combination and this will more clearer.

File: python/sgl_jax/test/test_kda_attention.py (hunk 7 above)
Comment (MokusMokun) [reply python/sgl_jax/test/test_kda_attention.py]: refactored

File: python/sgl_jax/test/test_kda_attention.py (hunk 1 above)
Comment (MokusMokun) [reply python/sgl_jax/test/test_kda_attention.py:23]: comment updated with more details
tp4dp4 e2e tested with MMLU-pro, reached desired baseline, check: https://github.com/sgl-project/sglang-jax/pull/1047#issuecomment-4437174216

File: python/sgl_jax/test/test_kda_attention_dp.py (hunk 11 above)
Comment (MokusMokun) [reply python/sgl_jax/test/test_kda_attention_dp.py:645]: added

File: python/sgl_jax/test/test_kda_attention_dp.py (hunk 12 above)
Comment (MokusMokun) [reply python/sgl_jax/test/test_kda_attention_dp.py:838]: added

File: python/sgl_jax/test/test_kda_attention_dp.py (hunk 10 above)
Comment (MokusMokun) [reply python/sgl_jax/test/test_kda_attention_dp.py:441]: refactored

File: python/sgl_jax/test/test_kda_attention_dp.py (hunk 13 above)
Comment (MokusMokun) [reply python/sgl_jax/test/test_kda_attention_dp.py:839]: refactored

File: python/sgl_jax/test/test_kda_attention.py (hunk 8 above)
Comment (MokusMokun) [reply python/sgl_jax/test/test_kda_attention.py:376]: refactored

File: python/sgl_jax/test/test_kda_attention_dp.py (hunk 14 above)
Comment (MokusMokun) [reply python/sgl_jax/test/test_kda_attention_dp.py:854]: tested tp2dp4 and tp4dp2 with `tpu-v6e-2x4`
check https://github.com/sgl-project/sglang-jax/issues/1038#issuecomment-4421798683

File: python/sgl_jax/test/test_kda_attention.py (hunk 9 above)
Comment (MokusMokun) [reply python/sgl_jax/test/test_kda_attention.py:616]: refactored

File: python/sgl_jax/srt/layers/radix_linear_attention.py [hunk 15]
```diff
@@ -69,7 +69,7 @@ def __call__(
             v=v,
             layer=self,
             forward_batch=forward_batch,
-            pool=recurrent_state_pool,
+            recurrent_state_pool=recurrent_state_pool,
```
Comment (hlFu) [inline python/sgl_jax/srt/layers/radix_linear_attention.py]: This needs to align with HybridLinearAttnBackend signature? However, HybridLinearAttnBackend still uses pool

File: python/sgl_jax/srt/layers/attention/linear/kda_backend.py [hunk 16]
```diff
@@ -0,0 +1,402 @@
...
+        )(conv_buffer, recurrent_indices)
+
+        has_initial_state = self.forward_metadata.has_initial_state
+        if has_initial_state is not None:
+            ssm = jnp.where(has_initial_state[:, None, None, None], ssm, 0.0)
+            conv = jnp.where(has_initial_state[:, None, None], conv, 0.0)
+
+        return ssm, conv
+
+    def set_ssm_state(self, recurrent_state_pool, layer_id, recurrent_indices, new_recurrent):
```
Comment (hlFu) [inline python/sgl_jax/srt/layers/attention/linear/kda_backend.py:176]: Can you consider move this inside gated delta kernel? I see 2 benefits of doing that:
1. This adds an additional shard_map which will cause an extra HBM round-trip as it's fusion boundary.
2. Maybe this set can be embedded into the pallas kernel where the get happens so to avoid reloading>
Same as the set_conv_state below

File: python/sgl_jax/srt/kernels/kda/kda.py [hunk 17]
```diff
@@ -0,0 +1,1278 @@
+# Adapted from https://github.com/primatrix/pallas-kernel (rev 3c691ad3)
```
Comment (hlFu) [inline python/sgl_jax/srt/kernels/kda/kda.py:1]: Looks like the adaption source is not accessible. Can you point to a public reference of the kernel implementation if possible?

File: python/sgl_jax/srt/layers/radix_linear_attention.py (hunk 15 above)
Comment (aolemila) [reply python/sgl_jax/srt/layers/radix_linear_attention.py]: Put your modification in this PR. Refer to https://github.com/MokusMokun/sglang-jax/pull/9/changes @MokusMokun

File: python/sgl_jax/srt/layers/attention/linear/kda_backend.py (hunk 16 above)
Comment (aolemila) [reply python/sgl_jax/srt/layers/attention/linear/kda_backend.py:176]: @pathfinder-pf may fuse this set op into kernel. For this PR, we consider the implementation and accuracy alignment. The performance will be considered in next phase.

File: python/sgl_jax/srt/kernels/kda/kda.py (hunk 17 above)
Comment (aolemila) [reply python/sgl_jax/srt/kernels/kda/kda.py:1]: Sorry. Currently this repo is private, and we have tested the e2e accuracy. So temporarily we use it. Could you recommend a public kernel implementation which has aligned the accuracy?

File: python/sgl_jax/srt/layers/radix_linear_attention.py (hunk 15 above)
Comment (MokusMokun) [reply python/sgl_jax/srt/layers/radix_linear_attention.py]: aligned

**Review by aolemila — APPROVED** [review]
